### PR TITLE
Wrapper for Crawly.Spider behaviour

### PR DESCRIPTION
**Pull request by oltarasenko**
1. Extend Crawly.Spider with spider_closed callback
2. Use macro to simplify the Crawly.Spider definitions

### CODE REVIEW

**Comment by Ziinc** [comment]
I agree with the `use Crawly.Spider` macro, I think it would simplify boilerplate a lot.

I don't agree with the use of the `spider_closed/1` callback. This implies that there will be a callback for every spider-related event. I feel that this will give rise to a lot of boilerplate, and a lot of copy-pasting of code. There is also no access to the engine/spider state.

As I mentioned in the [issue](https://github.com/oltarasenko/crawly/issues/74#issuecomment-606708681), I lean towards simply passing a function for the engine to call  through the settings. It is very extensible and already allows for spider-level overrides.

Some benefits from this approach:
- The callback is fired asynchronously, which would be beneficial for other types of events
- can define a default callback, override on a per-spider basis as necessary
- can separate this callback from spider code (so cleaner spider code)
- less maintenance (less callbacks on the spider)
- easier to understand (less callbacks that may behave differently, this one is standardized to receive engine state and spider state)
- more extensible (the task firing can be abstracted to a util function, and adding support for other events would just be a few lines of code)

For example:
```elixir
# in config.exs
config :crawly,
  on_spider_close: &MyApp.send_email_notification/2

# in MyApp.MySpider
def override_settings() do
    [on_spider_close: &MyApp.handle_update_spider_dashboard/2 ]
end

# in Engine.close_spider
def stop_spider(spider_name, reason \\ :ignore) do
  callback = Crawly.Utils.get_settings(spider_name, :on_spider_close)

  # start the callback that we don't care about
   if callback do
     Supervisor.start_link([
        {Task, fn -> callback.(engine_state, spider_state) end}
      ], strategy: :one_for_one)

   end

  # close the spider
  GenServer.call(__MODULE__, {:stop_spider, spider_name})
end

```

**Comment by oltarasenko** [comment]
>  don't agree with the use of the spider_closed/1 callback. This implies that there will be a callback for every spider-related event. I feel that this will give rise to a lot of boilerplate, and a lot of copy-pasting of code. There is also no access to the engine/spider state.
As I mentioned in the issue, I lean towards simply passing a function for the engine to call through the settings. It is very extensible and already allows for spider-level overrides.

I understand your concern (I was playing with both ways of declaring things before submitting the PR), and at this point, my arguments are the following.

**Custom settings way of defining callbacks have the following drawbacks**
 - custom_settings - is a KW list which allows overriding settings. **Semantically spider callbacks are not settings**. Even more, there is no global representation of the callbacks.
 - if we define for example spider_closed callback via settings, it would mean that our settings are more complex, now it's not only simple overrides of globally defined settings but also something on the top of it. 
- behavior itself will not be that clear: As well as callbacks are not defined by behavior, we would literally have to depend on conventions, which could be quite error-prone, when people will **do typos** and will not understand why something is not being called, and there will not be hints from the compiler. 
- finally, the Crawly internal code could be more simple. We would not have to do function_exported?/3 all the time. We can safely call something, expecting it to be defined default with the help of __using__ macro.

As far as I see the only benefit of custom_settings based approach is the fact that all spiders written for older versions of Crawly can be easily migrated without a switch to __using__ macro.

> Some benefits from this approach:

> The callback is fired asynchronously, which would be beneficial for other types of events
> can define a default callback, override on a per-spider basis as necessary

I think this can be easily achieved with the callback-based approach. We can force the macro to generate a function that will read the config by default.

> can separate this callback from spider code (so cleaner spider code)

Same as above, at this point more strict interface seems to be quite beneficial

> less maintenance (less callbacks on the spider)

Callbacks are not executed in the context of the spider. So technically it's just a place where the code is stored.

> more extensible (the task firing can be abstracted to a util function, and adding support for other events would just be a few lines of code)

We still can do it. E.g. the global setting might call a util function similarly to what you're suggesting.

Stating all the arguments above I would still prefer to go with the callback-based approach

**Comment by Ziinc** [comment]
> I think this can be easily achieved with the callback-based approach. We can force the macro to generate a function that will read the config by default

But wouldn't this be going back one full circle to defining the callback through settings? 

I agree that the callback isn't settings. But neither is it spider code. It shouldn't be living in the spider as well. These type of event-triggered callbacks are side-effects and are not related to data-parsing, which is the context of a spider module. 

From a user story perspective:
1. As a dev, I want to do something when this spider ends -- spider specific
2. As a dev, I want to do something when any of my spider ends -- not spider specific
3. As a dev, I want to do something when all my spiders end. -- not spider specific

The settings-approach allows for addressing of scenarios 1,2,3. While the macro callback approach only addresses scenario 1.

Relying on the macro to define the callbacks in advance also makes it a breaking change, and requires internal code to check if the callback is defined anyway.

Why have 2 different approaches to configuring a Spider when there already is a unified approach to configure it? In the end, if there's going to be a default setting on the config, it can be customized with `override_settings` and makes the callback moot.

**Comment by Ziinc** [comment]
I feel that the KW approach forces the dev to define the callback outside of the spider scope, which would be more appropriate for scenarios 2 and 3.

**Comment by Ziinc** [comment]
I also think the typo and compiler concerns can be rectified through debug logging.

I think the crux of the issue here is:
1. where should the callback code be defined, in spider or elsewhere
    - if in spider, then macro callback is an ok approach
    - if not in spider, then define elsewhere and point to it in settings.
2. How is this type of configuration (event callbacks) going to be extended in the future.
   - macro approach means 1 callback per event
   - KW setting approach means one new setting per event
3. How to specify a default (i.e. global) callback?

**Comment by oltarasenko** [comment]
> But wouldn't this be going back one full circle to defining the callback through settings?
Maybe not, if the settings are clearly defined and the callbacks code is autogenerated with a macro

> I agree that the callback isn't settings. But neither is it spider code. It shouldn't be living in the 
 spider as well. These type of event-triggered callbacks are side-effects and are not related to data- parsing, which is the context of a spider module.

Mmm.. Yes, you're right. Maybe they should not be a part of the spider at all. Let me ask for an advise from Scrapy core team members.

> Relying on the macro to define the callbacks in advance also makes it a breaking change,
Yes, it's a breaking change. 

> and requires internal code to check if the callback is defined anyway.
well, this is going to be handled by the compiler, who will return a warning stating that you have not defined required callbacks.

> Why have 2 different approaches to configuring a Spider when there already is a unified approach to configure it? In the end, if there's going to be a default setting on the config, it can be customized with override_settings and makes the callback moot.

Well, but from the other side, I see some similarity between parse_item/1 and spider_closed/1 or item_scraped/2 [item scraped would be even more specific to a given spider, e.g. it would allow to make some very specific post-processings]

**Comment by oltarasenko** [comment]
@Ziinc Actually after some conversations, I have decided to switch to the idea of the settings_based approach.

Please have another glance on the PR.

File: lib/crawly/spider.ex [hunk 1]
```diff
@@ -13,6 +13,9 @@ defmodule Crawly.Spider do
       provide custom spider specific settings. Should define a list with custom
       settings and their values. These values will take precedence over the
       global settings defined in the config.
+  5. `spider_closed/1` an optional callback which can be used in order to add
```
Comment (Ziinc) [inline lib/crawly/spider.ex:16]: Not used anymore, can remove

File: lib/crawly/spider.ex [hunk 2]
```diff
@@ -23,6 +26,17 @@ defmodule Crawly.Spider do
               Crawly.ParsedItem.t()
 
   @callback override_settings() :: Crawly.Settings.t()
+  @callback spider_closed(reason) :: :ok
```
Comment (Ziinc) [inline lib/crawly/spider.ex:29]: Not used

File: lib/crawly/spider.ex [hunk 3]
```diff
@@ -23,6 +26,17 @@ defmodule Crawly.Spider do
...
+            when reason: atom()
 
-  @optional_callbacks override_settings: 0
+  defmacro __using__(_opts) do
+    quote do
+      @behaviour Crawly.Spider
+
+      def override_settings(), do: []
+      def spider_closed(_reason), do: :ok
```
Comment (Ziinc) [inline lib/crawly/spider.ex:37]: Not used

File: lib/crawly/settings.ex [hunk 4]
```diff
@@ -43,6 +43,7 @@ defmodule Crawly.Settings do
           # Defines retries
           retry: retry(),
           middlewares: [middleware()],
-          pipelines: [pipeline()]
+          pipelines: [pipeline()],
+          on_spider_closed_callback: function()
```
Comment (Ziinc) [inline lib/crawly/settings.ex:47]: Do you think it'll be good to nest it under an `events` or `callbacks` key? For organisational purpose

Comment (oltarasenko) [reply lib/crawly/settings.ex:47]: I am not quite sure, from one side I don't want to have nested sections, as the configuration will be more complex. Lets maybe keep it as it is for now, and see if we need to tweak later.

**Pull request merged**